### PR TITLE
currency.Coin extension

### LINKED ISSUE #1306: Change following to currency.Coin
- [x] txn.Value
- [x] txn.Fee
- [x] delegatePool.Balance
- [x] delegatePool.Reward
- [x] minersc.SimpleNode.TotalStaked
- [x] stakepool/stakepool.go/spUpdate.Reward
- [x] Treat txn.Value as SAS, and notify the related callers about this changing.

### PR DESCRIPTION
**Pull request by ssardana08**
## Fixes
- txn Value converted to currency.Coin
- txn Fees converted to currency.Coin
- delegatepool fees, rewards changed to currency.coin
- stakepool fees, rewards changed to currency.coin

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test: https://github.com/0chain/system_test/pull/293
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | 0chain_issue_1306  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | issue_1306  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | 0chain_issue_1306  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | issue_1306  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | test-min-submit-100  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | issue_1306  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

File: code/go/0chain.net/smartcontract/dbs/event/delegate_pool.go [hunk 1]
```diff
@@ -63,14 +65,35 @@ func (sp *DelegatePool) exists(edb *EventDb) (bool, error) {
...
+			}
+		} else {
+			dpu.Updates["reward"], err = currency.MinusInt64(dp.Reward, reward)
+			if err != nil {
+				return err
+			}
+			dpu.Updates["total_reward"], err = currency.MinusInt64(dp.TotalReward, reward)
+			if err != nil {
+				return err
+			}
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/dbs/event/delegate_pool.go]: The previous way to code this seems simpler. So could be:

```go
if reward < 0 && provider is blobber {
  dpu.Update["total_penalty"] = currency.AddInt64(dp.TotalPenalty, reward)
} else {
  dpu.Updates["reward"], err = currency.AddInt64(dp.Reward, reward)
  if err != nil {
    return err
  }
  dpu.Updates["total_reward"], err = currency.AddInt64(dp.TotalReward, reward)
  if err != nil {
    return err
   
   }

```

File: code/go/0chain.net/smartcontract/dbs/event/delegate_pool.go [hunk 2]
```diff
@@ -63,14 +65,35 @@ func (sp *DelegatePool) exists(edb *EventDb) (bool, error) {
...
+		if err != nil {
+			return err
+		}
+		dpu.Updates["total_reward"], err = currency.AddInt64(dp.TotalReward, reward)
+		if err != nil {
+			return err
+		}
 	}
 	if err := edb.updateDelegatePool(*dpu); err != nil {
 		return nil
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/dbs/event/delegate_pool.go]: seems error is not returned

File: code/go/0chain.net/smartcontract/storagesc/config.go [hunk 3]
```diff
@@ -382,7 +386,10 @@ func getConfiguredConfig() (conf *Config, err error) {
 		pfx + "writepool.max_lock_period")
 	// stake pool
 	conf.StakePool = new(stakePoolConfig)
-	conf.StakePool.MinLock = int64(scc.GetFloat64(pfx+"stakepool.min_lock") * 1e10)
+	conf.StakePool.MinLock, err = currency.ParseZCN(scc.GetFloat64(pfx + "stakepool.min_lock"))
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/config.go:389]: The MinLock here use ZCN unit, but zcnsc use SAS unit, we need to unify the unit, but can do it later.

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/storagesc/config.go (hunk 3 above)
Comment (ssardana08) [reply code/go/0chain.net/smartcontract/storagesc/config.go:389]: maybe we will have all these config changed to zcn, sas and be parsed

**Review by peterlimg — APPROVED** [review]
lgtm

**Pull request merged**